packages/lib/src/components/UPI/UPI.tsx

```diff
@@ -0,0 +1,108 @@
+    public static type = 'upi';
+
+    constructor(props: any) {
+        super(props);
+        this.handleGenerateQrCodeClick = this.handleGenerateQrCodeClick.bind(this);
+    }
+
+    public get isValid(): boolean {
+        return !!this.state.isValid;
+    }
+
+    public formatData(): UpiPaymentData {
+        const { isQrCodeFlow, virtualPaymentAddress } = this.state.data;
+        return {
+            paymentMethod: {
+                type: isQrCodeFlow ? UPI_QR : UPI_COLLECT,
+                ...(virtualPaymentAddress && { virtualPaymentAddress })
+            }
+        };
+    }
+
+    public setStatus(status: UIElementStatus): this {
+        this.componentRef?.setStatus(status, this.state.data.isQrCodeFlow);
+        return this;
+    }
+
+    private handleGenerateQrCodeClick(): void {
+        this.setState({ data: { isQrCodeFlow: true }, valid: {}, errors: {}, isValid: true });
+        this.submit();
+    }
+
+    private renderContent(type: string): h.JSX.Element {
+        switch (type) {
+            case 'qrCode':
+                return (
+                    <QRLoader
+                        ref={ref => {
+                            this.componentRef = ref;
+                        }}
+                        {...this.props}
+                        type={UPI_QR}
+                        brandLogo={this.props.brandLogo || this.icon}
+                        onComplete={this.onComplete}
+                        introduction={this.props.i18n.get('upi.qrCodeWaitingMessage')}
+                        countdownTime={5}
+                    />
+                );
+            case 'await':
+                return (
+                    <Await
+                        ref={ref => {
+                            this.componentRef = ref;
+                        }}
+                        clientKey={this.props.clientKey}
+                        paymentData={this.props.paymentData}
+                        onError={this.handleError}
+                        onComplete={this.onComplete}
+                        brandLogo={this.icon}
+                        type={UPI_COLLECT}
+                        messageText={this.props.i18n.get('upi.vpaWaitingMessage')}
+                        awaitText={this.props.i18n.get('await.waitForConfirmation')}
+                        showCountdownTimer
+                        countdownTime={5}
+                    />
+                );
+            default:
+                return (
+                    <UPIComponent
+                        ref={(ref: RefObject<typeof UPIComponent>) => {
+                            this.componentRef = ref;
+                        }}
+                        payButton={this.payButton}
+                        onChange={this.setState}
+                        onSubmit={this.submit}
+                        onGenerateQrCodeClick={this.handleGenerateQrCodeClick}
+                    />
+                );
+        }
+    }
+
+    public render(): h.JSX.Element {
```

Comment:
Do we need to type the render function?
Also why h.JSX.Element? Is this what we should use? I did something for i18n an was using Preact.VNode

Reply:
We don't necessarily need to type as we aren't using strict mode. In reality we don't need to type anything 😅 . But I don't see why not type it - if we turn on the strict typescript mode at some point, we will have less issues with untyped code.. and I believe that makes sense to push for more strict typing in the future

About `h.JSX.Element` vs `preact.VNode` , I am not fully sure which one is recommended to go for. I see few usages of both in the codebase. I used the `h.JSX.Element` as it is the type inferred by the Typescript compiler in any 'render' function that returns JSX.

Reply:
Well, I don't think render is going to cause issues in strict mode since we extend Component for the component.
I did search and I did not find consensus too. I will change some of my code to be `JSX.Element` since it seems to be a bit more framework agnostic.